### ISSUE TITLE
Build failure:`or-tools` on `x86_64-darwin`: "no member named 'aligned_alloc' in the global namespace"

### ISSUE THREAD

**Issue by donn** [NixOS/nixpkgs#272156]
### Steps To Reproduce

1. Save this expression to `reproducible.nix`:
```
{
    pkgs ? import (fetchTarball "https://github.com/NixOS/nixpkgs/archive/8078ceb2777d790d3fbc53589ed3753532185d77.tar.gz") {
        overlays = [
            (new: old: {
                cbc = old.cbc.overrideAttrs(finalAttrs: previousAttrs: {
                    configureFlags = previousAttrs.configureFlags ++ ["CXXFLAGS=-std=c++14"]; 
                });
            })
        ];
    }
}:

pkgs.mkShell {
    buildInputs = [pkgs.or-tools];
}
```
  * 8078ceb2777d790d3fbc53589ed3753532185d77 is the HEAD of release-23.11 at the time of writing
  * Overlay for `cbc` is related to https://github.com/NixOS/nixpkgs/issues/234710, specifically Clang 16's default of C++17:
    * `register` compiler hint causes an error (can be worked around with `-Wno-register`)
    * `bind2nd` has been removed in C++17
2. `nix-shell --option system x86_64-darwin --extra-platforms x86_64-darwin reproducible.nix`

### Build log

https://gist.github.com/donn/4ace429bbdfe43e3707cddfbc0c13697

### Additional context

The same expression runs fine on `aarch64-darwin` without a build failure.

I ran this using Rosetta. I don't have immediate access to an Intel Mac atm but there may be a tiny chance that this is a Rosetta bug.

### Notify maintainers

@andersk 

### Metadata

```console
 - system: `"aarch64-darwin"`
 - host os: `Darwin 23.1.0, macOS 14.1.1`
 - multi-user?: `yes`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.18.1`
 - channels(root): `"nixpkgs"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixpkgs`
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

**Comment by wegank** [comment]
`aligned_alloc` isn't available until macOS 10.15, while x86_64-darwin defaults to SDK 10.12.

I guess something like `stdenv = if stdenv.isDarwin then overrideSDK stdenv "11.0" else stdenv;` in an additional override might help.

**Comment by donn** [comment]
Thank you so much for the help- it compiles and all the tests pass with the macOS 11 SDK.

I'll keep this issue open until https://github.com/NixOS/nixpkgs/pull/272247 is merged.
